### ISSUE TITLE
Function called with different types

### ISSUE THREAD

**Issue by hashbackup** [seq-lang/seq#94]
Shouldn't this code work?  I thought that if a function was called with different arg types, the compiler would generate two different functions, one for each arg type.  Instead, it looks like the compiler is complaining because it is inferring the type of x to be both int and str.
```
[root@hbseq ~]# cat argtype2.py
def fn(x):
    if x == 'abc':
        print 'str'
    elif x == 1:
        print 'int'
    else:
        print 'other'

fn('abc')
fn(1)
fn([1,2])
[root@hbseq ~]# python2 argtype2.py
str
int
other
[root@hbseq ~]# ./s argtype2
+ seqc -o argtype2.bc argtype2.seq
seqc: /lib64/libtinfo.so.5: no version information available (required by /root/.seq/libseq.so)
argtype2.py:4:9: error: unsupported operand type(s) for ==: 'str' and 'int'

**Comment by markhend** [comment]
Interesting example. It appears the compiler infers x is a str and so then balks when checking that the types are the same on either side of `==` in `x == 1`. I played with your example a bit. This is about as close as I could get:
```
def fn(x):
    y = str(x)
    if y == 'abc':
        print 'str'
    elif y == '1':
        print 'int'
    else:
        print 'other
```

**Comment by hashbackup** [comment]
First I tried this:
```
[root@hbseq ~]# cat argtype.seq
def fn(x):
    if type(x) == str:
        print 'str'
    elif type(x) == int:
        print 'int'
    else:
        print 'other'

fn('abc')
fn(1)
fn([1,2])

[root@hbseq ~]# ./s argtype
seqc: /lib64/libtinfo.so.5: no version information available (required by /root/.seq/libseq.so)
argtype.seq:2:7: error: parsing error
```
I guess maybe type() isn't implemented yet?

I could be wrong and probably am, but my intuition is that Seq would have a better chance of large adoption and more contributions if it was "a compiled Python with genomic extensions" vs "a genomics language similar to Python".

**Comment by arshajii** [comment]
Yes, this is because the type of `x` becomes `str` or `int` depending on the argument type, which is what produces the error. **However,** I think this is something we can actually solve even in a statically-typed context. We can potentially check at compile time when two things of different types are being compared like that (and there is no corresponding `__eq__` magic) and remove that branch completely. So I think we can actually solve this case (and it's quite a common pattern in Python). We're in the process of reworking our parser and type checker at the moment, and I'll see if we can do something like this.

I would definitely like to frame Seq as a compiled Python with genomic extensions, especially since it's basically a general-purpose language at its core. It's just that's a much steeper hill to climb (need to add many more libraries, features, etc.), and we don't have all that much manpower at the moment. But I agree there :)

**Comment by hashbackup** [comment]
Using the 2nd example, there would be no type mismatch on the == operator.  I don't know how your type inference works, but it seems that it looks at a function definition to infer argument types without considering the actual arguments on a function *call*.  If you know the actual arguments' type, that should take precedence over whatever is going on inside the function.

Another way to look at it is if the function call was inlined.  In that case, fn('abc') would compile to the single statement print 'str' and similarly with the other two calls.

However, I'm sure it's much more involved / complex than I understand! :-)

**Comment by inumanag** [comment]
Mostly resolved in #132 .

We added support for compile-time conditionals, so you can do this to simulate a function that returns multiple types:

```
def foo(x):
   if isinstance(x, int):
      print 'int'; return x
   elif isinstance(x, str):
      print 'str'; return x
```
